README
ncurses module for Qore
version 0.2 by David Nichols

Requires qore 0.7.1 and ncurses or SYSV curses headers and libraries to build.

Documentation TBD

This module provides the "Panel" and "Window" classes along with many functions and constants - see test/ncurses.q and examples/hanio.q and examples/worms.q for more information. 

note that this module is still experimental due to the fact that I'm not sure if it's possible to safely enable threading without putting a big lock around every curses call.  Basically threading and curses do not mix.
if your ncurses is in a non-standard location, set the NCURSES_DIR environment variable before running configure.  Also can be built with Solaris curses.

test script: test/ncurses.q
example scripts: examples/

BUILD AND INSTALL
-----------------

configure
make
sudo make install

The configure script will find out where your qore module directory is found and set this for the install directory.

BUILD FROM SVN
--------------

reconf.sh
configure
make
sudo make install


ChangeLog
2010-06-15 18:49  david_nichols

	* docs/ncurses-module-doc.html: updated tikiwiki and added updated
	  color scheme

2010-05-12 11:49  david_nichols

	* configure.ac: solaris and sun studio compile and warning
	  fixes/enhancements

2010-05-11 14:45  david_nichols

	* reconf.sh: compile fixes for solaris

2010-04-28 10:15  david_nichols

	* examples/worm.q: updates for new qore API and to address new
	  warnings in test scripts

2009-12-09 12:58  david_nichols

	* configure.ac: fixed all modules to use decimal comparisons with
	  the qore version string instead of octal so that they will still
	  compile with qore v0.8.0 and up...

2009-11-22 06:03  david_nichols

	* configure.ac: updated configure to assume 64-bit build if
	  host_cpu = "amd64"

2009-10-30 08:22  david_nichols

	* configure.ac: updated all module configure scripts to use gnutar
	  on darwin to avoid picking up OSX resource files in source
	  releases

2009-09-21 11:07  david_nichols

	* configure.ac: updated module configure scripts to select 64-bit
	  target mode by default on snow leopard i386

2009-09-10 06:49  david_nichols

	* configure.ac: updated configure.ac to properly compile in 32-bit
	  mode on OSX snow leopard (10.6)

2009-07-20 13:16  david_nichols

	* configure.ac, qore-ncurses-module.spec: added opensuse 11.2
	  detection to spec file, fixed rhel detection in spec files,
	  updated configure.ac to work with older versions of autotools

2009-06-08 08:38  david_nichols

	* qore-ncurses-module.spec: fixed qore module dir detection

2009-06-02 05:24  david_nichols

	* qore-ncurses-module.spec: fixed typo in sles version detection

2009-01-08 18:43  david_nichols

	* qore-ncurses-module.spec: simplified spec files a little

2009-01-08 15:13  david_nichols

	* qore-ncurses-module.spec: commented-out additional "Source0" line
	  so that packages will build on Fedora 10 (OBS)

2009-01-08 14:44  david_nichols

	* reconf.sh: updated reconf.sh to look for glibtoolize as well as
	  libtoolize

2009-01-07 12:51  david_nichols

	* Makefile.am, RELEASE-NOTES, configure.ac, ltmain.sh,
	  qore-ncurses-module.spec, reconf.sh: updated reconf.sh to use
	  libtoolize to generate ltmain.sh; updated to compile when panel.h
	  is under the ncurses subdirectory (ex: OpenSuse 11.1)

2009-01-03 22:51  david_nichols

	* src/ncurses.cc: cosmetic updates

2008-11-03 14:14  david_nichols

	* ChangeLog, src/QC_Panel.cc, src/QC_Window.cc,
	  src/ncurses-module.h, src/ncurses.cc: make module version reflect
	  package version

2008-10-30 17:20  david_nichols

	* qore-ncurses-module.spec: various fixes

2008-10-30 10:42  david_nichols

	* Makefile.am, README, RELEASE-NOTES, configure.ac,
	  qore-ncurses-module.spec, src/Makefile.am: udpated to qoer 0.7.1
	  module handling

2008-10-23 06:30  david_nichols

	* configure.ac: updated to set 64-bit compile mode by default on
	  itanium with g++ and aCC

2008-10-08 20:16  david_nichols

	* qore-ncurses-module.spec, src/Makefile.am: spec file and module
	  build fixes using prefix=/usr and make install
	  DESTDIR=$RPM_BUILD_ROOT

2008-10-06 20:27  david_nichols

	* ChangeLog: updated changelog

2008-09-09 09:11  david_nichols

	* docs/ncurses-module-doc.html: updated docs

2008-09-09 07:04  david_nichols

	* ChangeLog, qore-ncurses-module.spec: updated ChangeLog, fixed
	  license tag in spec file

2008-09-09 07:01  david_nichols

	* Makefile.am, docs, docs/ncurses-module-doc.html,
	  qore-ncurses-module.spec: added module documentation

2008-09-03 12:41  david_nichols

	* src/Makefile.am: build fixes

2008-09-03 12:11  david_nichols

	* COPYING: updated license for modules to LGPL (*not* GPL)

2008-09-03 12:07  david_nichols

	* README: udpated READMEs

2008-09-03 09:23  david_nichols

	* Makefile.am, configure.ac, src/Makefile.am: build fixes

2008-09-03 09:19  david_nichols

	* ., AUTHORS, COPYING, ChangeLog, Makefile.am, NEWS, README,
	  configure.ac, examples, examples/hanoi.q, examples/worm.q,
	  ltmain.sh, m4, m4/acx_pthread.m4, qore-ncurses-module.spec,
	  reconf.sh, src, src/Makefile.am, src/QC_Panel.cc, src/QC_Panel.h,
	  src/QC_Window.cc, src/QC_Window.h, src/ncurses-module.h,
	  src/ncurses.cc, src/single-compilation-unit.cc, test,
	  test/ncurses.q: initial import as independent project



src/ncurses-module.h
/*
  modules/ncurses/ncurses.h

  Qore Programming Language

  Copyright 2003 - 2008 David Nichols

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

#ifndef _QORE_NCURSES_MODULE_H

#define _QORE_NCURSES_MODULE_H

#include "../config.h"

#include <qore/Qore.h>

QoreStringNode *ncurses_module_init();
void ncurses_module_ns_init(class QoreNamespace *rns, class QoreNamespace *qns);
void ncurses_module_delete();

#include <qore/QoreThreadLock.h>

class q_nc_init_class {
   private:
      bool initialized;
      class QoreThreadLock l;

      void init_intern();

   public:
      inline q_nc_init_class() : initialized(false) {}
      void close();
      inline void init()
      {
	 // the "init" variable can only be set once, so we check
	 // it outside the lock, and only if it's then do we grab the lock and 
	 // recheck and initialize if necessary.  This should avoid unnecessary
	 // locking and unlocking (cache syncs, etc) the vast majority of the 
	 // time and still remain thread-safe
	 if (initialized)
	    return;
	 init_intern();
      }
};

extern class q_nc_init_class q_nc_init;


#endif // _QORE_NCURSES_MODULE_H


src/QC_Panel.cc

/*
  QC_Panel.cc

  Qore Programming Language

  Copyright 2003 - 2008 David Nichols

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

#include "ncurses-module.h"

#include "QC_Panel.h"

qore_classid_t CID_PANEL;

//class QoreThreadLock nc_panel_update_lock;

void PC_constructor(class QoreObject *self, const QoreListNode *params, ExceptionSink *xsink) {
   const AbstractQoreNode *p0, *p1, *p2, *p3;
   p0 = get_param(params, 0);
   p1 = get_param(params, 1);
   p2 = get_param(params, 2);
   p3 = get_param(params, 3);
   int lines   = p0 ? p0->getAsInt() : 0;
   int columns = p1 ? p1->getAsInt() : 0;
   int y       = p2 ? p2->getAsInt() : 0;
   int x       = p3 ? p3->getAsInt() : 0;

   // ensure that the curses library has been initialized
   q_nc_init.init();
   class Panel *p = new Panel(lines, columns, y, x, xsink);
   if (xsink->isException())
      p->deref();
   self->setPrivate(CID_PANEL, p);
}

static void PC_copy(class QoreObject *self, class QoreObject *old, class Panel *p, class ExceptionSink *xsink)
{
   xsink->raiseException("PANEL-COPY-ERROR", "copying Panel objects is currently unsupported");
}

AbstractQoreNode *PC_keypad(class QoreObject *self, class Panel *p, const QoreListNode *params, ExceptionSink *xsink)
{
   const AbstractQoreNode *p0 = get_param(params, 0);
   bool b = p0 ? p0->getAsBool() : false;

   AbstractQoreNode *rv;
   int rc = p->keypad(b);
   rv = new QoreBigIntNode(rc);
   return rv;
}

AbstractQoreNode *PC_addstr(class QoreObject *self, class Panel *p, const QoreListNode *params, ExceptionSink *xsink)
{
   AbstractQoreNode *rv;
   int rc;
   const QoreStringNode *p0 = test_string_param(params, 0);
   if (p0)
      rc = p->qaddstr(p0->getBuffer());
   else
      rc = 0;

   rv = new QoreBigIntNode(rc);
   return rv;
}

AbstractQoreNode *PC_mvaddstr(class QoreObject *self, class Panel *p, const QoreListNode *params, ExceptionSink *xsink)
{
   AbstractQoreNode *rv;
   int rc;
   const QoreStringNode *p2 = test_string_param(params, 2);
   if (p2)
   {
      const AbstractQoreNode *p0 = get_param(params, 0);
      const AbstractQoreNode *p1 = get_param(params, 1);
      rc = p->qmvaddstr(p0 ? p0->getAsInt() : 0, p1 ? p1->getAsInt() : 0, p2->getBuffer());
   }
   else
      rc = 0;

   rv = new QoreBigIntNode(rc);
   return rv;
}

AbstractQoreNode *PC_printw(class QoreObject *self, class Panel *p, const QoreListNode *params, ExceptionSink *xsink)
{
   QoreStringNodeHolder str(q_sprintf(params, 0, 0, xsink));
   return new QoreBigIntNode(p->qaddstr(str->getBuffer()));
}

AbstractQoreNode *PC_mvprintw(class QoreObject *self, class Panel *p, const QoreListNode *params, ExceptionSink *xsink)
{
   QoreStringNodeHolder str(q_sprintf(params, 0, 2, xsink));
   const AbstractQoreNode *p0 = get_param(params, 0);
   const AbstractQoreNode *p1 = get_param(params, 1);

   return new QoreBigIntNode(p->qmvaddstr(p0 ? p0->getAsInt() : 0, p1 ? p1->getAsInt() : 0, str->getBuffer()));
}

AbstractQoreNode *PC_refresh(class QoreObject *self, class Panel *p, const QoreListNode *params, ExceptionSink *xsink)
{
   p->qrefresh();
   return NULL;
}

AbstractQoreNode *PC_update(class QoreObject *self, class Panel *p, const QoreListNode *params, ExceptionSink *xsink)
{
   p->update();
   return NULL;
}

AbstractQoreNode *PC_getch(class QoreObject *self, class Panel *p, const QoreListNode *params, ExceptionSink *xsink)
{
   AbstractQoreNode *rv;
   rv = new QoreBigIntNode(p->qgetch());
   return rv;
}

AbstractQoreNode *PC_border(class QoreObject *self, class Panel *p, const QoreListNode *params, ExceptionSink *xsink)
{
   AbstractQoreNode *rv;
   const AbstractQoreNode *p0, *p1, *p2, *p3, *p4, *p5, *p6, *p7;
   p0 = get_param(params, 0);
   p1 = get_param(params, 1);
   p2 = get_param(params, 2);
   p3 = get_param(params, 3);
   p4 = get_param(params, 4);
   p5 = get_param(params, 5);
   p6 = get_param(params, 6);
   p7 = get_param(params, 7);
   int ls = p0 ? p0->getAsInt() : 0;
   int rs = p1 ? p1->getAsInt() : 0;
   int ts = p2 ? p2->getAsInt() : 0;
   int bs = p3 ? p3->getAsInt() : 0;
   int tl = p4 ? p4->getAsInt() : 0;
   int tr = p5 ? p5->getAsInt() : 0;
   int bl = p6 ? p6->getAsInt() : 0;
   int br = p7 ? p7->getAsInt() : 0;
   
   rv = new QoreBigIntNode(p->qborder(ls, rs, ts, bs, tl, tr, bl, br));
   return rv;
}

AbstractQoreNode *PC_setColor(class QoreObject *self, class Panel *p, const QoreListNode *params, ExceptionSink *xsink)
{
   AbstractQoreNode *rv;
   const AbstractQoreNode *p0 = get_param(params, 0);
   const AbstractQoreNode *p1 = get_param(params, 1);
   rv = new QoreBigIntNode(p->setColor(p0 ? p0->getAsInt() : COLOR_WHITE,
					p1 ? p1->getAsInt() : COLOR_BLACK));
   return rv;
}

AbstractQoreNode *PC_setBackgroundColor(class QoreObject *self, class Panel *p, const QoreListNode *params, ExceptionSink *xsink)
{
   AbstractQoreNode *rv;
   const AbstractQoreNode *p0 = get_param(params, 0);
   rv = new QoreBigIntNode(p->setBackgroundColor(p0 ? p0->getAsInt() : COLOR_BLACK));
   return rv;
}

AbstractQoreNode *PC_setForegroundColor(class QoreObject *self, class Panel *p, const QoreListNode *params, ExceptionSink *xsink)
{
   AbstractQoreNode *rv;
   const AbstractQoreNode *p0 = get_param(params, 0);
   rv = new QoreBigIntNode(p->setForegroundColor(p0 ? p0->getAsInt() : COLOR_WHITE));
   return rv;
}

AbstractQoreNode *PC_show(class QoreObject *self, class Panel *p, const QoreListNode *params, ExceptionSink *xsink)
{
   return new QoreBigIntNode(p->show());
}

AbstractQoreNode *PC_hide(class QoreObject *self, class Panel *p, const QoreListNode *params, ExceptionSink *xsink)
{
   return new QoreBigIntNode(p->hide());
}

AbstractQoreNode *PC_movePanel(class QoreObject *self, class Panel *p, const QoreListNode *params, ExceptionSink *xsink)
{
   const AbstractQoreNode *p0 = get_param(params, 0);
   const AbstractQoreNode *p1 = get_param(params, 1);
   return new QoreBigIntNode(p->movePanel(p0 ? p0->getAsInt() : 0,
					   p1 ? p1->getAsInt() : 0));
}

AbstractQoreNode *PC_move(class QoreObject *self, class Panel *p, const QoreListNode *params, ExceptionSink *xsink)
{
   const AbstractQoreNode *p0 = get_param(params, 0);
   const AbstractQoreNode *p1 = get_param(params, 1);
   return new QoreBigIntNode(p->qmove(p0 ? p0->getAsInt() : 0,
				       p1 ? p1->getAsInt() : 0));
}

AbstractQoreNode *PC_top(class QoreObject *self, class Panel *p, const QoreListNode *params, ExceptionSink *xsink)
{
   return new QoreBigIntNode(p->top());
}

AbstractQoreNode *PC_bottom(class QoreObject *self, class Panel *p, const QoreListNode *params, ExceptionSink *xsink)
{
   return new QoreBigIntNode(p->bottom());
}

AbstractQoreNode *PC_getLines(class QoreObject *self, class Panel *p, const QoreListNode *params, ExceptionSink *xsink)
{
   return new QoreBigIntNode(p->getLines());
}

AbstractQoreNode *PC_getColumns(class QoreObject *self, class Panel *p, const QoreListNode *params, ExceptionSink *xsink)
{
   return new QoreBigIntNode(p->getColumns());
}

#ifdef HAVE_WRESIZE
AbstractQoreNode *PC_resize(class QoreObject *self, class Panel *p, const QoreListNode *params, ExceptionSink *xsink)
{
   const AbstractQoreNode *p0 = get_param(params, 0);
   const AbstractQoreNode *p1 = get_param(params, 1);
   return new QoreBigIntNode(p->resize(p0 ? p0->getAsInt() : 0,
					p1 ? p1->getAsInt() : 0));
}
#endif

AbstractQoreNode *PC_getY(class QoreObject *self, class Panel *p, const QoreListNode *params, ExceptionSink *xsink)
{
   return new QoreBigIntNode(p->getY());
}

AbstractQoreNode *PC_getX(class QoreObject *self, class Panel *p, const QoreListNode *params, ExceptionSink *xsink)
{
   return new QoreBigIntNode(p->getX());
}

AbstractQoreNode *PC_getBegY(class QoreObject *self, class Panel *p, const QoreListNode *params, ExceptionSink *xsink)
{
   return new QoreBigIntNode(p->getBegY());
}

AbstractQoreNode *PC_getBegX(class QoreObject *self, class Panel *p, const QoreListNode *params, ExceptionSink *xsink)
{
   return new QoreBigIntNode(p->getBegX());
}

AbstractQoreNode *PC_attrset(class QoreObject *self, class Panel *p, const QoreListNode *params, ExceptionSink *xsink)
{
   const AbstractQoreNode *p0 = get_param(params, 0);
   return new QoreBigIntNode(p->qattrset(p0 ? p0->getAsInt() : 0));
}

AbstractQoreNode *PC_attron(class QoreObject *self, class Panel *p, const QoreListNode *params, ExceptionSink *xsink)
{
   const AbstractQoreNode *p0 = get_param(params, 0);
   return new QoreBigIntNode(p->qattron(p0 ? p0->getAsInt() : 0));
}

AbstractQoreNode *PC_attroff(class QoreObject *self, class Panel *p, const QoreListNode *params, ExceptionSink *xsink)
{
   const AbstractQoreNode *p0 = get_param(params, 0);
   return new QoreBigIntNode(p->qattroff(p0 ? p0->getAsInt() : 0));
}

AbstractQoreNode *PC_scrollok(class QoreObject *self, class Panel *p, const QoreListNode *params, ExceptionSink *xsink)
{
   const AbstractQoreNode *p0 = get_param(params, 0);
   return new QoreBigIntNode(p->scrollok(p0 ? p0->getAsBool() : 0));
}


AbstractQoreNode *PC_idlok(class QoreObject *self, class Panel *p, const QoreListNode *params, ExceptionSink *xsink)
{
   const AbstractQoreNode *p0 = get_param(params, 0);
   return new QoreBigIntNode(p->idlok(p0 ? p0->getAsBool() : 0));
}

AbstractQoreNode *PC_clearok(class QoreObject *self, class Panel *p, const QoreListNode *params, ExceptionSink *xsink)
{
   const AbstractQoreNode *p0 = get_param(params, 0);
   return new QoreBigIntNode(p->clearok(p0 ? p0->getAsBool() : 0));
}

AbstractQoreNode *PC_idcok(class QoreObject *self, class Panel *p, const QoreListNode *params, ExceptionSink *xsink)
{
   const AbstractQoreNode *p0 = get_param(params, 0);
   p->idcok(p0 ? p0->getAsBool() : 0);
   return NULL;
}

AbstractQoreNode *PC_immedok(class QoreObject *self, class Panel *p, const QoreListNode *params, ExceptionSink *xsink)
{
   const AbstractQoreNode *p0 = get_param(params, 0);
   p->immedok(p0 ? p0->getAsBool() : 0);
   return NULL;
}

AbstractQoreNode *PC_erase(class QoreObject *self, class Panel *p, const QoreListNode *params, ExceptionSink *xsink)
{
   p->qerase();
   return NULL;
}

AbstractQoreNode *PC_clear(class QoreObject *self, class Panel *p, const QoreListNode *params, ExceptionSink *xsink)
{
   p->qclear();
   return NULL;
}

AbstractQoreNode *PC_setscrreg(class QoreObject *self, class Panel *p, const QoreListNode *params, ExceptionSink *xsink)
{
   const AbstractQoreNode *p0 = get_param(params, 0);
   const AbstractQoreNode *p1 = get_param(params, 1);
   return new QoreBigIntNode(p->qsetscrreg(p0 ? p0->getAsInt() : 0, p1 ? p1->getAsInt() : 0));
}

AbstractQoreNode *PC_redraw(class QoreObject *self, class Panel *p, const QoreListNode *params, ExceptionSink *xsink)
{
   return new QoreBigIntNode(p->redraw());
}

AbstractQoreNode *PC_scroll(class QoreObject *self, class Panel *p, const QoreListNode *params, ExceptionSink *xsink)
{
   return new QoreBigIntNode(p->qscroll());
}

AbstractQoreNode *PC_scrl(class QoreObject *self, class Panel *p, const QoreListNode *params, ExceptionSink *xsink)
{
   const AbstractQoreNode *p0 = get_param(params, 0);
   return new QoreBigIntNode(p->qscrl(p0 ? p0->getAsInt() : 0));
}

AbstractQoreNode *PC_hline(class QoreObject *self, class Panel *p, const QoreListNode *params, ExceptionSink *xsink)
{
   const AbstractQoreNode *p0 = get_param(params, 0);
   const AbstractQoreNode *p1 = get_param(params, 1);
   return new QoreBigIntNode(p->qhline(p0 ? p0->getAsInt() : 0, p1 ? p1->getAsInt() : 0));
}

AbstractQoreNode *PC_vline(class QoreObject *self, class Panel *p, const QoreListNode *params, ExceptionSink *xsink)
{
   const AbstractQoreNode *p0 = get_param(params, 0);
   const AbstractQoreNode *p1 = get_param(params, 1);
   return new QoreBigIntNode(p->qvline(p0 ? p0->getAsInt() : 0, p1 ? p1->getAsInt() : 0));
}

AbstractQoreNode *PC_mvhline(class QoreObject *self, class Panel *p, const QoreListNode *params, ExceptionSink *xsink)
{
   const AbstractQoreNode *p0 = get_param(params, 0);
   const AbstractQoreNode *p1 = get_param(params, 1);
   const AbstractQoreNode *p2 = get_param(params, 2);
   const AbstractQoreNode *p3 = get_param(params, 3);
   return new QoreBigIntNode(p->qmvhline(p0 ? p0->getAsInt() : 0, 
					  p1 ? p1->getAsInt() : 0,
					  p2 ? p2->getAsInt() : 0,
					  p3 ? p3->getAsInt() : 0));
}

AbstractQoreNode *PC_mvvline(class QoreObject *self, class Panel *p, const QoreListNode *params, ExceptionSink *xsink)
{
   const AbstractQoreNode *p0 = get_param(params, 0);
   const AbstractQoreNode *p1 = get_param(params, 1);
   const AbstractQoreNode *p2 = get_param(params, 2);
   const AbstractQoreNode *p3 = get_param(params, 3);
   return new QoreBigIntNode(p->qmvvline(p0 ? p0->getAsInt() : 0, 
					  p1 ? p1->getAsInt() : 0,
					  p2 ? p2->getAsInt() : 0,
					  p3 ? p3->getAsInt() : 0));
}

AbstractQoreNode *PC_addch(class QoreObject *self, class Panel *p, const QoreListNode *params, ExceptionSink *xsink)
{
   const AbstractQoreNode *p0 = get_param(params, 0);
   if (!is_nothing(p0) && (p0->getType() != NT_STRING || (reinterpret_cast<const QoreStringNode *>(p0))->strlen()))
      return new QoreBigIntNode(p->qaddch(getChar(p0)));
   return 0;
}

AbstractQoreNode *PC_mvaddch(class QoreObject *self, class Panel *p, const QoreListNode *params, ExceptionSink *xsink)
{
   const AbstractQoreNode *p0 = get_param(params, 0);
   const AbstractQoreNode *p1 = get_param(params, 1);
   const AbstractQoreNode *p2 = get_param(params, 2);
   if (!is_nothing(p2) && (p2->getType() != NT_STRING || (reinterpret_cast<const QoreStringNode *>(p2))->strlen()))
      return new QoreBigIntNode(p->qmvaddch(p0 ? p0->getAsInt() : 0,
					    p1 ? p1->getAsInt() : 0,
					    getChar(p2)));
   return 0;
}

AbstractQoreNode *PC_clrtoeol(class QoreObject *self, class Panel *p, const QoreListNode *params, ExceptionSink *xsink)
{
   p->qclrtoeol();
   return NULL;
}

AbstractQoreNode *PC_clrtobot(class QoreObject *self, class Panel *p, const QoreListNode *params, ExceptionSink *xsink)
{
   p->qclrtobot();
   return NULL;
}

class QoreClass *initPanelClass()
{
   QORE_TRACE("initPanelClass()");

   class QoreClass *QC_PANEL = new QoreClass("Panel", QDOM_TERMINAL_IO);
   CID_PANEL = QC_PANEL->getID();
   QC_PANEL->setConstructor(PC_constructor);
   QC_PANEL->setCopy((q_copy_t)PC_copy);
   QC_PANEL->addMethod("keypad",             (q_method_t)PC_keypad);
   QC_PANEL->addMethod("mvaddstr",           (q_method_t)PC_mvaddstr);
   QC_PANEL->addMethod("mvprintw",           (q_method_t)PC_mvprintw);
   QC_PANEL->addMethod("addstr",             (q_method_t)PC_addstr);
   QC_PANEL->addMethod("printw",             (q_method_t)PC_printw);
   QC_PANEL->addMethod("refresh",            (q_method_t)PC_refresh);
   QC_PANEL->addMethod("update",             (q_method_t)PC_update);
   QC_PANEL->addMethod("getch",              (q_method_t)PC_getch);
   QC_PANEL->addMethod("border",             (q_method_t)PC_border);
   QC_PANEL->addMethod("setBackgroundColor", (q_method_t)PC_setBackgroundColor);
   QC_PANEL->addMethod("setForegroundColor", (q_method_t)PC_setForegroundColor);
   QC_PANEL->addMethod("setColor",           (q_method_t)PC_setColor);
   QC_PANEL->addMethod("getLines",           (q_method_t)PC_getLines);
   QC_PANEL->addMethod("getColumns",         (q_method_t)PC_getColumns);

   QC_PANEL->addMethod("show",               (q_method_t)PC_show);
   QC_PANEL->addMethod("hide",               (q_method_t)PC_hide);
   QC_PANEL->addMethod("top",                (q_method_t)PC_top);
   QC_PANEL->addMethod("bottom",             (q_method_t)PC_bottom);

#ifdef HAVE_WRESIZE
   QC_PANEL->addMethod("resize",             (q_method_t)PC_resize);
#endif
   QC_PANEL->addMethod("getY",               (q_method_t)PC_getY);
   QC_PANEL->addMethod("getX",               (q_method_t)PC_getX);
   QC_PANEL->addMethod("getBegY",            (q_method_t)PC_getBegY);
   QC_PANEL->addMethod("getBegX",            (q_method_t)PC_getBegX);

   QC_PANEL->addMethod("attrset",            (q_method_t)PC_attrset);
   QC_PANEL->addMethod("attron",             (q_method_t)PC_attron);
   QC_PANEL->addMethod("attroff",            (q_method_t)PC_attroff);
   QC_PANEL->addMethod("movePanel",          (q_method_t)PC_movePanel);
   QC_PANEL->addMethod("move",               (q_method_t)PC_move);
   QC_PANEL->addMethod("scrollok",           (q_method_t)PC_scrollok);
   QC_PANEL->addMethod("idlok",              (q_method_t)PC_idlok);
   QC_PANEL->addMethod("clearok",            (q_method_t)PC_clearok);
   QC_PANEL->addMethod("idcok",              (q_method_t)PC_idcok);
   QC_PANEL->addMethod("immedok",            (q_method_t)PC_immedok);
   QC_PANEL->addMethod("erase",              (q_method_t)PC_erase);
   QC_PANEL->addMethod("clear",              (q_method_t)PC_clear);
   QC_PANEL->addMethod("redraw",             (q_method_t)PC_redraw);
   QC_PANEL->addMethod("scroll",             (q_method_t)PC_scroll);
   QC_PANEL->addMethod("scrl",               (q_method_t)PC_scrl);
   QC_PANEL->addMethod("setscrreg",          (q_method_t)PC_setscrreg);
   QC_PANEL->addMethod("hline",              (q_method_t)PC_hline);
   QC_PANEL->addMethod("vline",              (q_method_t)PC_vline);
   QC_PANEL->addMethod("mvhline",            (q_method_t)PC_mvhline);
   QC_PANEL->addMethod("mvvline",            (q_method_t)PC_mvvline);
   QC_PANEL->addMethod("addch",              (q_method_t)PC_addch);
   QC_PANEL->addMethod("mvaddch",            (q_method_t)PC_mvaddch);

   QC_PANEL->addMethod("clrtoeol",           (q_method_t)PC_clrtoeol);
   QC_PANEL->addMethod("clrtobot",           (q_method_t)PC_clrtobot);


   return QC_PANEL;
}
